pages/villageHQ.php

```diff
@@ -442,32 +448,204 @@ function villageHQ() {
+                        $lesson_modifier = SenseiManager::getLessonModifier($player->taijutsu_skill, $player->exp, $lesson_sensei['taijutsu_skill'], $lesson_sensei['exp'], $lesson_sensei['specialization'] == 'taijutsu_skill' ? true : false);
+                        break;
+                    case 'genjutsu_skill':
+                        $lesson_modifier = SenseiManager::getLessonModifier($player->genjutsu_skill, $player->exp, $lesson_sensei['genjutsu_skill'], $lesson_sensei['exp'], $lesson_sensei['specialization'] == 'genjutsu_skill' ? true : false);
+                        break;
+                    case 'bloodline_skill':
+						if ($player->bloodline_id == $lesson_sensei['bloodline_id']) {
+							$lesson_modifier = round(SenseiManager::getLessonModifier($player->bloodline_skill, $player->exp, $lesson_sensei['bloodline_skill'], $lesson_sensei['exp'], true) * 100 - 100, 2);
+						} else {
+							throw new Exception('Invalid stat!');
+                        }
+                        break;
+                    case 'speed':
+                        $lesson_modifier = SenseiManager::getLessonModifier($player->speed, $player->exp, $lesson_sensei['speed'], $lesson_sensei['exp'], $lesson_sensei['specialization'] == 'speed' ? true : false);
+                        break;
+                    case 'cast_speed':
+                        $lesson_modifier = SenseiManager::getLessonModifier($player->cast_speed, $player->exp, $lesson_sensei['cast_speed'], $lesson_sensei['exp'], $lesson_sensei['specialization'] == 'cast_speed' ? true : false);
+                        break;
+                    default:
+                        throw new Exception('Invalid stat!');
+                }
+
+				// Verify training has minimum increase
+				if ($lesson_modifier <= 1) {
+					throw new Exception('Invalid stat!');
+                }
+
+				// Verify player can afford training
+				$player->subtractMoney($lesson_cost, "Paid " . $lesson_cost . "&yen; for lessons.");
+
+                // Add yen to sensei
+                $sensei_player = User::loadFromId($system, $lesson_sensei['sensei_id']);
+                $sensei_player->loadData(User::UPDATE_NOTHING);
+                $sensei_player->addMoney($lesson_cost / 5, "Earned " . $lesson_cost / 5 . "&yen; for lessons.");
+                $sensei_player->updateData();
+
+				// Set training for player
+				$stat_train_gain = 4 + ($player->rank_num * 4);
+				$stat_long_train_gain = $stat_train_gain * 2.25;
+                $stat_extended_train_gain = $stat_train_gain * 12;
+				$stat_long_train_gain *= $player->forbidden_seal->long_training_gains;
+                $stat_extended_train_gain = round($stat_extended_train_gain * $player->forbidden_seal->extended_training_gains);
+                $stat_train_gain += $system->TRAIN_BOOST;
+                $stat_long_train_gain += $system->LONG_TRAIN_BOOST;
+                $stat_extended_train_gain += ($system->LONG_TRAIN_BOOST * 5);
```

Comment:
Tbh I would really prefer we centralize these in the PR that they are introduced (currently, nowhere outside of `training.php` attempts to define these values) rather than introducing them and then cleaning up later. Would be okay if it's in total shell of a `TrainingManager` class which is only used (for now) for exposing these values to wherever needs them.

But I am okay with that being an immediate fast follow

Reply:
Moved to TrainingManager